src/main/java/uk/gov/pay/api/auth/AccountAuthenticator.java

```diff
@@ -55,6 +57,7 @@ public AccountAuthenticator(Client client, PublicApiConfig configuration) {
     private Optional<Account> readAccountFromResponse(Response response) {
         JsonNode responseEntity = response.readEntity(JsonNode.class);
         String accountId = responseEntity.get("account_id").asText();
+        MDC.put(GATEWAY_ACCOUNT_ID, accountId);
```

Comment:
This might not get invoked all the time. `AccountAuthenticator` result is cached by Dropwizard, so AuthFilter might not invoke `AccountAuthenticator. readAccountFromResponse` if object still exists in cache (but not exactly sure of the behaviour)
https://github.com/alphagov/pay-publicapi/blob/master/src/main/java/uk/gov/pay/api/app/PublicApi.java#L120

We can probably create MdcFilter as a ContainerRequestFilter (which we can set to be invoked after Authentication) and it can both put and clear `gateway_account_id`.

Reply:
Its a great point about the caching and therefor we cannot set the value here. As discussed I don't think we can use a `ContainerRequestFilter` to tidy up and call `MDC.clean()` since it has no means to be called back when the request is complete. Looking at `LoggingFilter` it will remove the request-id mdc value if present before setting it. This seems a pragmatic approach but could lead to some confusion if anything is called prior to that filter as it would have a stale value. Perhaps a separate container filter to add it with `USER` priority and another servlet filter at the head of the chain to clean all mdc values...

Reply:
Is it still required here ?

Reply:
I'd just realised that I'd left this in, I've removed and pushed.